### PR TITLE
fix clang headers to support HIP c kernels and to support OpenMP .  T…

### PR DESCRIPTION
**Pull request by gregrodgers**
…his depricates the autoheader __clang_openmp_runtime_wrapper.h .  Both nvptx and amdgcn will use the autoheader opemp_wrappers/__clang_openmp_device-functions.h

With this PR we stop using libm device lib for OpenMP.

### CODE REVIEW

File: clang/lib/Headers/__clang_hip_libdevice_declares.h [hunk 1]
```diff
@@ -126,10 +123,16 @@ __device__ __attribute__((const)) float __ocml_div_rte_f32(float, float);
 __device__ __attribute__((const)) float __ocml_div_rtn_f32(float, float);
 __device__ __attribute__((const)) float __ocml_div_rtp_f32(float, float);
 __device__ __attribute__((const)) float __ocml_div_rtz_f32(float, float);
-__device__ __attribute__((const)) float __ocml_sqrt_rte_f32(float, float);
-__device__ __attribute__((const)) float __ocml_sqrt_rtn_f32(float, float);
-__device__ __attribute__((const)) float __ocml_sqrt_rtp_f32(float, float);
-__device__ __attribute__((const)) float __ocml_sqrt_rtz_f32(float, float);
+
+//__device__ __attribute__((const)) float __ocml_sqrt_rte_f32(float, float);
```
Comment (JonChesterfield) [inline clang/lib/Headers/__clang_hip_libdevice_declares.h:127]: This looks like a breaking change - do we need to drop the second argument?

File: clang/lib/Headers/__clang_hip_complex_builtins.h [hunk 2]
```diff
@@ -0,0 +1,208 @@
+/*===-- __clang_hip_complex_builtins - HIP impls of runtime complex fns ---===
+ *
+ * Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See https://llvm.org/LICENSE.txt for license information.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ *===-----------------------------------------------------------------------===
+ */
+
+#ifndef __CLANG_HIP_COMPLEX_BUILTINS
```
Comment (JonChesterfield) [inline clang/lib/Headers/__clang_hip_complex_builtins.h:10]: could be worth moving this out of extras as a second patch to make the line delta look less intimidating to review (on gerrit or on phab)

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -1170,7 +1170,8 @@ void Clang::AddPreprocessingOptions(Compilation &C, const JobAction &JA,
   // If we are offloading to a target via OpenMP we need to include the
   // openmp_wrappers folder which contains alternative system headers.
   if (JA.isDeviceOffloading(Action::OFK_OpenMP) &&
-      getToolChain().getTriple().isNVPTX()){
+      (getToolChain().getTriple().isNVPTX() ||
+       (getToolChain().getTriple().getArch() == llvm::Triple::amdgcn))) {
```
Comment (JonChesterfield) [inline clang/lib/Driver/ToolChains/Clang.cpp:1174]: iirc there's an isAMDGCN() function available now which does that ^

File: clang/lib/Headers/__clang_hip_libdevice_declares.h [hunk 4]
```diff
@@ -294,8 +298,13 @@ __device__ __attribute__((const)) _Float16 __ocml_trunc_f16(_Float16);
 typedef _Float16 __2f16 __attribute__((ext_vector_type(2)));
 typedef short __2i16 __attribute__((ext_vector_type(2)));
 
+#if defined(__cplusplus)
```
Comment (JonChesterfield) [inline clang/lib/Headers/__clang_hip_libdevice_declares.h:301]: This looks sketchy. How about, somewhere near the top of the file, 
```
#if !defined(__cplusplus)
#include <stdbool.h>
#endif
```

File: clang/lib/Headers/__clang_hip_libdevice_declares.h [hunk 5]
```diff
@@ -310,17 +319,21 @@ __device__ __attribute__((const)) __2i16 __ocml_isnan_2f16(__2f16);
 __device__ __attribute__((pure)) __2f16 __ocml_log_2f16(__2f16);
 __device__ __attribute__((pure)) __2f16 __ocml_log10_2f16(__2f16);
 __device__ __attribute__((pure)) __2f16 __ocml_log2_2f16(__2f16);
+#if defined(__cplusplus)
```
Comment (JonChesterfield) [inline clang/lib/Headers/__clang_hip_libdevice_declares.h:322]: This compiles fine as C if the return uses return (__2f16){...}, don't need to disable it entirely. amd-stg-open already has that patch.

File: clang/lib/Headers/__clang_hip_runtime_wrapper.h [hunk 6]
```diff
@@ -18,40 +18,54 @@
 
 #if __HIP__
 
+#if defined(__cplusplus)
 #include <cmath>
 #include <cstdlib>
 #include <stdlib.h>
+#define _NULLPTR nullptr
```
Comment (JonChesterfield) [inline clang/lib/Headers/__clang_hip_runtime_wrapper.h:25]: This seems messy. We can return NULL in C++ and it'll do the right thing, or we can return 0 and that'll do the right thing too.

**Review by JonChesterfield — COMMENTED** [review]
Bigger change than anticipated, sadly. Thanks for writing this.

Rebasing on #158 (e.g. by applying that to amd-stg-openmp first) will remove some of the line noise and will let this apply directly to amd-stg-open, without manual intervention.

**Comment by ronlieb** [comment]
seems to be not too invasive
interested to here if Sam is good with it all.

File: clang/lib/Driver/ToolChains/CommonArgs.cpp [hunk 7]
```diff
@@ -1628,6 +1628,10 @@ bool tools::GetSDLFromAOB(Compilation &C, const Driver &D, const Tool &T,
   return FoundAOB;
 }
 
+//  Lets clean up this overloading!
```
Comment (aaronenyeshi) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:1631]: These comments won't be needed when upstreaming.

File: clang/lib/Driver/ToolChains/CommonArgs.cpp [hunk 8]
```diff
@@ -1681,9 +1688,9 @@ void tools::AddStaticDeviceLibs(Compilation *C, const Tool *T,
   SmallVector<std::string, 16> SDL_Names;
   for (std::string SDL_Name : DriverArgs.getAllArgValues(options::OPT_l)) {
     // No Device specific SDL for these libs: omp,cudart,m,gcc,gcc_s,pthread
-    if ( SDL_Name != "omp" && SDL_Name != "cudart" && SDL_Name != "m" &&
-         SDL_Name != "gcc" && SDL_Name != "gcc_s" && SDL_Name != "pthread" &&
-         SDL_Name != "hip_hcc" ) {
+    if (SDL_Name != "omp" && SDL_Name != "cudart" && SDL_Name != "m" &&
+        SDL_Name != "gcc" && SDL_Name != "gcc_s" && SDL_Name != "pthread" &&
+        SDL_Name != "hip_hcc" && SDL_Name != "m") {
```
Comment (aaronenyeshi) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:1693]: `SDL_Name != "m"` is already used in line 1691. `if (SDL_Name != "omp" && SDL_Name != "cudart" && SDL_Name != "m"`

File: clang/lib/Headers/__clang_hip_math.h [hunk 9]
```diff
@@ -365,821 +452,331 @@ inline void sincospif(float __x, float *__sinptr, float *__cosptr) {
...
-  double __tmp;
-  *__sinptr = __ocml_sincospi_f64(
-      __x, (__attribute__((address_space(5))) double *)&__tmp);
-  *__cosptr = __tmp;
-}
+float __saturatef(float x) { return (x < 0) ? 0 : ((x > 1) ? 1 : x); }
+
 __DEVICE__
-inline double sinh(double __x) { return __ocml_sinh_f64(__x); }
+double sinh(double x) { return __ocml_sinh_f64(x); }
```
Comment (aaronenyeshi) [inline clang/lib/Headers/__clang_hip_math.h:631]: where are the rest of the double functions? starting from acos to sincospi?

File: clang/lib/Headers/__clang_hip_math.h [hunk 10]
```diff
@@ -365,821 +452,331 @@ inline void sincospif(float __x, float *__sinptr, float *__cosptr) {
...
+float __powf(float x, float y) { return __ocml_pow_f32(x, y); }
+
 __DEVICE__
-inline void sincospi(double __x, double *__sinptr, double *__cosptr) {
-  double __tmp;
-  *__sinptr = __ocml_sincospi_f64(
-      __x, (__attribute__((address_space(5))) double *)&__tmp);
-  *__cosptr = __tmp;
-}
+float __saturatef(float x) { return (x < 0) ? 0 : ((x > 1) ? 1 : x); }
```
Comment (aaronenyeshi) [inline clang/lib/Headers/__clang_hip_math.h:628]: looks like the float intrinsics got cut short too. Missing __sincosf, __sinf, __cosf.

File: clang/lib/Headers/__clang_hip_runtime_wrapper.h [hunk 11]
```diff
@@ -18,40 +18,54 @@
...
+#include <stddef.h>
+#endif
 
 #define __host__ __attribute__((host))
 #define __device__ __attribute__((device))
 #define __global__ __attribute__((global))
 #define __shared__ __attribute__((shared))
 #define __constant__ __attribute__((constant))
+#define __forceinline__ __attribute__((always_inline))
+#define __private
```
Comment (aaronenyeshi) [inline clang/lib/Headers/__clang_hip_runtime_wrapper.h:37]: is this define missing `_attribute__((address_space(5)))`

**Review by aaronenyeshi — CHANGES_REQUESTED** [review]
I'm fixing merge conflict if this was to merge into amd-stg-open and upstream master.

File: clang/lib/Headers/__clang_hip_math.h [hunk 12]
```diff
@@ -1,362 +1,449 @@
...
-  uint64_t __r = 0;
-  while (__tagp) {
-    char __tmp = *__tagp;
+#define USED __attribute__((used))
 
-    if (__tmp >= '0' && __tmp <= '7')
-      __r = (__r * 8u) + __tmp - '0';
+// These are defined in cmath so skip them if cplusplus
+#ifndef __cplusplus
+__DEVICE__ inline uint64_t __make_mantissa_base8(const char *tagp) {
```
Comment (yxsamliu) [inline clang/lib/Headers/__clang_hip_math.h:48]: clang requires all headers uses prefix function parameter names with `__`

File: clang/lib/Headers/openmp_wrappers/complex [hunk 13]
```diff
@@ -15,11 +15,45 @@
...
+#define __ARCHTYPES__ nvptx,nvptx64
+#endif
+
+#ifdef __AMDGCN__
+#include <__clang_hip_libdevice_declares.h>
+#define __HIP__
+#endif
 #include <__clang_cuda_complex_builtins.h>
+#define __ARCHTYPES__ amdgcn
 #endif
```
Comment (aaronenyeshi) [inline clang/lib/Headers/openmp_wrappers/complex:30]: which ifdef/ifndef is this endif for?

**Review by aaronenyeshi — CHANGES_REQUESTED** [review]

File: clang/lib/Headers/openmp_wrappers/complex (hunk 13 above)
Comment (aaronenyeshi) [reply clang/lib/Headers/openmp_wrappers/complex:30]: this seems to end the header file.

File: clang/lib/Headers/__clang_hip_complex_builtins.h (hunk 2 above)
Comment (gregrodgers) [reply clang/lib/Headers/__clang_hip_complex_builtins.h]: This is deleted.  Both amdgcn and nvptx use __clang_cuda_complex_builtins.h

File: clang/lib/Driver/ToolChains/CommonArgs.cpp (hunk 7 above)
Comment (gregrodgers) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:1631]: This was a reminder to me that we don't need three methods for AddStaticDeviceLibs.

**Comment by gregrodgers** [comment]
I believe I have addressed all the comments.  I am ready to merge this for 11.9-1.  We should be able to delete libm from aomp-extras now.

**Comment by ronlieb** [comment]
have you addressed the test failures ethan observed? and these that i observed ?
do you want us to retest with your latest patch 
They are all FORTRAN TESTS

Add to the list of tests breaking
Build errors:
Nekbone
SPEC ACCEL 550.pmd(peak; CE), 553.pclvrleaf(peak; CE), 556.psp(peak; CE), 563.pswim(peak; CE)
Spec hpc2021 619.clvleaf_s 635.weather_s/

**Review by ronlieb — CHANGES_REQUESTED** [review]
please address testing issues from ethan and i

**Comment by ronlieb** [comment]
The following fortrran applocation fail because these symbols are unresolved with this patch applied to the build 

sqrt, exp, sin, cos, pow

File: clang/lib/Headers/openmp_wrappers/complex [hunk 14]
```diff
@@ -15,11 +15,47 @@
...
+
+#ifdef __AMDGCN__
+#include <__clang_hip_libdevice_declares.h>
+#define __HIP__
+#define __ARCHTYPES__ amdgcn
+#endif  //__AMDGCN__
+
 #include <__clang_cuda_complex_builtins.h>
-#endif
+#endif //__CLANG_OPENMP_COMPLEX__
```
Comment (aaronenyeshi) [inline clang/lib/Headers/openmp_wrappers/complex:31]: If we end the header file here, the below code will not be used. Is that intended?

**Comment by estewart08** [comment]
Tested with gfx906:
**Smoke:**
math_modd of core dump memory access fault

**Hip-openmp:**
Quite a few failures in hip programs with openmp code, here is a snippet:
aomp_11.9-1/lib/clang/11.0.0/include/__clang_cuda_complex_builtins.h:138:10: error: no matching function for call to '__ocml_isinf_4'
        (_ISINFd(__ac) || _ISINFd(__bd) || _ISINFd(__ad) || _ISINFd(__bc))) {
aomp_11.9-1/lib/clang/11.0.0/include/__clang_cuda_complex_builtins.h:134:15: error: no matching function for call to '__ocml_copysign_f64'

**Apps:**
lulesh
error: call to 'pow' is ambiguous

snap
error: call to 'log2' is ambiguous

**Nekbone:**
lld: error: undefined symbol: sin
lld: error: undefined symbol: cos

**Comment by aaronenyeshi** [comment]
> The following fortrran applocation fail because these symbols are unresolved with this patch applied to the build
> 
> sqrt, exp, sin, cos, pow

this is related to my comment on __clang_hip_math.h. They are missing several float and double functions. Somehow the list got chopped in half.
https://github.com/ROCm-Developer-Tools/amd-llvm-project/pull/160#discussion_r495196419
https://github.com/ROCm-Developer-Tools/amd-llvm-project/pull/160#discussion_r495196088

**Comment by gregrodgers** [comment]
Sorry, we are not ready to merge.  The libm model provided a compiled library that made it easy for FORTRAN.  We cannot use header libraries with FORTRAN.

**Comment by gregrodgers** [comment]
The latest commits add all the missing c functions.   And a method to build a linkable bitcode from the header files __clang_hip_math.h or __clang_cuda_math.h.  .  In the aomp-extras repo there is a PR to compile a simple c file that is used to create the link libraries starting with libm- and ending in .bc.   

One last change is needed so that the FORTRAN driver links to the libm bc libraries.

**Comment by estewart08** [comment]
Tested new patchset with gfx906:

**Nekbone:***
Pass

**Smoke:**
math_modff: core dump memory access fault

**Hip-openmp:**
Quite a few failures in hip programs with openmp code, here is a snippet:
aomp_11.9-1/lib/clang/11.0.0/include/__clang_cuda_complex_builtins.h:138:10: error: no matching function for call to '__ocml_isinf_64'
(_ISINFd(__ac) || _ISINFd(__bd) || _ISINFd(__ad) || _ISINFd(__bc))) {
aomp_11.9-1/lib/clang/11.0.0/include/__clang_cuda_complex_builtins.h:134:15: error: no matching function for call to '__ocml_copysign_f64'

**Apps:**
lulesh
error: call to 'pow' is ambiguous

snap
error: call to 'log2' is ambiguous

hpgmg-mp4
error: conflicting types for 'norm'
aomp_11.9-1/lib/clang/11.0.0/include/__clang_hip_math.h:351:1: note: previous implicit declaration is here norm(int __dim,

**Comment by gregrodgers** [comment]
This PR is replaced by PR169
https://github.com/ROCm-Developer-Tools/amd-llvm-project/pull/169

**Pull request closed without merging**